### PR TITLE
Cardigann v4 (Newznab / Usenet Support) & More

### LINKED ISSUE #261: Switch Usenet Indexers and Torznab Implementations to YAML based
**Is your feature request related to a problem? Please describe.**

Switch Usenet and Torznab to YAML based for easier maintainability and allowing additional data

See additional context
**Describe the solution you'd like**

Switch from direct C# implementations to YAML based definitions.
**Describe alternatives you've considered**

**Additional context**

Blocking https://github.com/Prowlarr/Prowlarr/issues/259
Blocking https://github.com/Prowlarr/Prowlarr/issues/260

[AB#19](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/19)

### LINKED ISSUE #680: Add Headers to Cardigann Login and Search Blocks
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe

Title - we currently only support headers for Download; expand this to all the other blocks
- definition
- parser

### Describe the solution you'd like

title

### Describe alternatives you've considered

n/a

### Anything else?

n/a

[AB#2101](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/2101)

### LINKED ISSUE #782: Pull Jackett Commit - core: implement category_desc parser for cardigann indexers
https://github.com/Jackett/Jackett/commit/02e43bd6a9838562a14c6bd6ea10cfed5a293d59
fixup! https://github.com/Jackett/Jackett/commit/60d2c425e1038a28f54bda658a9f543f3a197a66

[AB#2372](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/2372)

### LINKED ISSUE #807: ShowRSS RSS feed does not work in combination with Sonarr
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

ShowRSS index created in Prowlarr and connected to Sonarr does not work because ShowRSS does not include the size of the torrent in its content, thus leading Sonarr to reject the item.

RSS Item:
```
<item>
<title>Batwoman 3x10 Toxic 1080p</title>
<link>magnet:?xt=urn:btih:F1FA71413974B93705CEB1D7A0611837A597C1F1&dn=Batwoman+S03E10+Toxic+1080p+AMZN+WEBRip+DDP5+1+x264+NOSiViD&tr=udp%3A%2F%2Ftracker.coppersurfer.tk%3A6969%2Fannounce&tr=udp%3A%2F%2Ftracker.opentrackr.org%3A1337%2Fannounce&tr=udp%3A%2F%2Fexplodie.org%3A6969&tr=udp%3A%2F%2F9.rarbg.me%3A2960&tr=udp%3A%2F%2F9.rarbg.to%3A2980</link>
<guid isPermaLink="false">8b6145e205f865773d4bf5d83b44f4ab5ace882c</guid>
<pubDate>Thu, 27 Jan 2022 20:10:29 +0000</pubDate>
<description>New episode: Batwoman S03E10 Toxic 1080p AMZN WEBRip DDP5 1 x264 NOSiViD. Link: <a href="magnet:?xt=urn:btih:F1FA71413974B93705CEB1D7A0611837A597C1F1&dn=Batwoman+S03E10+Toxic+1080p+AMZN+WEBRip+DDP5+1+x264+NOSiViD&tr=udp%3A%2F%2Ftracker.coppersurfer.tk%3A6969%2Fannounce&tr=udp%3A%2F%2Ftracker.opentrackr.org%3A1337%2Fannounce&tr=udp%3A%2F%2Fexplodie.org%3A6969&tr=udp%3A%2F%2F9.rarbg.me%3A2960&tr=udp%3A%2F%2F9.rarbg.to%3A2980">magnet:?xt=urn:btih:F1FA71413974B93705CEB1D7A0611837A597C1F1&dn=Batwoman+S03E10+Toxic+1080p+AMZN+WEBRip+DDP5+1+x264+NOSiViD&tr=udp%3A%2F%2Ftracker.coppersurfer.tk%3A6969%2Fannounce&tr=udp%3A%2F%2Ftracker.opentrackr.org%3A1337%2Fannounce&tr=udp%3A%2F%2Fexplodie.org%3A6969&tr=udp%3A%2F%2F9.rarbg.me%3A2960&tr=udp%3A%2F%2F9.rarbg.to%3A2980</a></description>
<tv:show_id>1206</tv:show_id>
<tv:external_id>37776</tv:external_id>
<tv:show_name>Batwoman</tv:show_name>
<tv:episode_id>156304</tv:episode_id>
<tv:raw_title>Batwoman S03E10 Toxic 1080p AMZN WEBRip DDP5 1 x264 NOSiViD</tv:raw_title>
<tv:info_hash>F1FA71413974B93705CEB1D7A0611837A597C1F1</tv:info_hash>
<enclosure url="magnet:?xt=urn:btih:F1FA71413974B93705CEB1D7A0611837A597C1F1&dn=Batwoman+S03E10+Toxic+1080p+AMZN+WEBRip+DDP5+1+x264+NOSiViD&tr=udp%3A%2F%2Ftracker.coppersurfer.tk%3A6969%2Fannounce&tr=udp%3A%2F%2Ftracker.opentrackr.org%3A1337%2Fannounce&tr=udp%3A%2F%2Fexplodie.org%3A6969&tr=udp%3A%2F%2F9.rarbg.me%3A2960&tr=udp%3A%2F%2F9.rarbg.to%3A2980" length="0" type="application/x-bittorrent"/>
</item>
```

Sonarr logs:
```
2022-01-28 07:33:25.9|Debug|DownloadDecisionMaker|Processing release 'Batwoman S03E10 Toxic 1080p AMZN WEBRip DDP5 1 x264 NOSiViD' from 'ShowRSS (Prowlarr)'
2022-01-28 07:33:25.9|Debug|Parser|Parsing string 'Batwoman S03E10 Toxic 1080p AMZN WEBRip DDP5 1 x264 NOSiViD'
2022-01-28 07:33:25.9|Debug|Parser|Episode Parsed. Batwoman - S03E10 
2022-01-28 07:33:25.9|Debug|Parser|Language parsed: English
2022-01-28 07:33:25.9|Debug|QualityParser|Trying to parse quality for Batwoman S03E10 Toxic 1080p AMZN WEBRip DDP5 1 x264 NOSiViD
2022-01-28 07:33:25.9|Debug|Parser|Quality parsed: WEBRip-1080p v1
2022-01-28 07:33:25.9|Debug|Parser|Release Group parsed: 
2022-01-28 07:33:25.9|Debug|AcceptableSizeSpecification|Beginning size check for: Batwoman S03E10 Toxic 1080p AMZN WEBRip DDP5 1 x264 NOSiViD
2022-01-28 07:33:25.9|Debug|AcceptableSizeSpecification|Item: Batwoman S03E10 Toxic 1080p AMZN WEBRip DDP5 1 x264 NOSiViD, Size: 512 is smaller than minimum allowed size (176160768 bytes for 42min), rejecting.
2022-01-28 07:33:25.9|Debug|AlreadyImportedSpecification|Performing already imported check on report
2022-01-28 07:33:25.9|Debug|AlreadyImportedSpecification|Skipping already imported check for episode without file
2022-01-28 07:33:25.9|Debug|LanguageSpecification|Checking if report meets language requirements. English
2022-01-28 07:33:25.9|Debug|MaximumSizeSpecification|Maximum size is not set.
2022-01-28 07:33:25.9|Debug|MinimumAgeSpecification|Not checking minimum age requirement for non-usenet report
2022-01-28 07:33:25.9|Debug|QualityAllowedByProfileSpecification|Checking if report meets quality requirements. WEBRip-1080p v1
2022-01-28 07:33:25.9|Debug|ReleaseRestrictionsSpecification|Checking if release meets restrictions: Batwoman S03E10 Toxic 1080p AMZN WEBRip DDP5 1 x264 NOSiViD
2022-01-28 07:33:25.9|Debug|ReleaseRestrictionsSpecification|[Batwoman S03E10 Toxic 1080p AMZN WEBRip DDP5 1 x264 NOSiViD] No restrictions apply, allowing
2022-01-28 07:33:25.9|Debug|RetentionSpecification|Not checking retention requirement for non-usenet report
2022-01-28 07:33:25.9|Debug|SceneMappingSpecification|No applicable scene mapping, skipping.
2022-01-28 07:33:25.9|Debug|DelaySpecification|QualityProfile does not require a waiting period before download for Torrent.
2022-01-28 07:33:25.9|Debug|HistorySpecification|Performing history status check on report
2022-01-28 07:33:25.9|Debug|HistorySpecification|Checking current status of episode [6425] in history
2022-01-28 07:33:25.9|Debug|DownloadDecisionMaker|Release rejected for the following reasons: [Permanent] 512.0 B is smaller than minimum allowed 168.0 MB (for 42min)
```

### Expected Behavior

The expected behavior would be for Prowlarr to use `Torrent RSS Feed` feature in Sonarr which contains Allow Zero Size (or to have this option in Prowlarr itself) to allow the correct usage of RSS feeds.
image

### Steps To Reproduce

_No response_

### Environment

```markdown
- OS: Synology
- Prowlarr: 0.2.0.1426
- Docker Install: No
- Using Reverse Proxy: No
- Browser: -
```

### What branch are you running?

Develop

### Trace Logs?

Can be found in `Current behavior`.

[AB#2509](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/2509)

### PR DESCRIPTION
**Pull request by bakerboy448**
# Cardigann Changes for v4

- Better SearchBlock and DownloadBlock header support and use
- Search Caps and Release Response
  - TMDBId
  - Genre
  - TraktID
  - CategoryDescr
- Protocol
- ignoreblankinputs (defined as SupportsEmptyInputs)
- allowdownloadredirect (defined as SupportsRedirect)
- SearchBlock Error
- DownloadBlock Error

#### Database Migration
NO

#### Description

- New: (Cardigann) Improved Search Logging
- New: (Cardigann) - Cardigann v4 Add Support for MapTrackerCatDescToNewznab
- New: (Cardigann) - Cardigann v4 Support for categorydesc
- New: (Cardigann) - Cardigann v4 Support Usenet (Newznab)
- New: (Cardigann) - Cardigann v4 Add Props Protocol and SupportsRedirect (allowdownloadredirect)
- New: (Cardigann) - Cardigann v4 Support for DownloadBlock Error
- New: (Cardigann) - Cardigann v4 Support for SearchBlock Error
- New: (Cardigann) - Cardigann v4 Add Prop SupportsEmptyInputs (allowblankinputs)
- New: (Cardigann) - Cardigann v4 Support for TMDbId, Genre, and TraktId
- New: (Cardigann) - Cardigann v4 Headers support for Download and Login Blocks
- Cardigann v4 - Purge default Newznab definitions and obsolete v4 indexers
- New: (Cardigann) - Cardigann v4 Support for Genre, Year, and TraktID
- Remove unnecessary usings
- Split Cardigann Release into a Release and a TorrentRelease

#### Cardigann Definitions

- [v4 YML Definitions](https://github.com/Prowlarr/Indexers/pull/79)

#### Code and Testing
- [X] Search Testing
- [X] Download Testing
- [X] Search Error Testing
- [x] Download Error Testing
- [x] Speed/Performance Testing
- [ ] Add support for one response, multiple categories or update the YML so we do not have duplicate category YML keys Defer to @Qstick thoughts as to how to handle that?
- [X] Obsolete all existing newznab
  - [X] Discuss how to do this. Defer to @Qstick 
- [ ] Convert BinSearch to YML
- [x]  Purge-Obsolete Newznab
- [ ] Convert newznab pre-definitions to YML
  - [ ] Waiting on @RobinDadswell to assist with finishing the Powershell Newznab Caps response => YML definition Script
  - [x] categories should be validated to the source code if we can https://github.com/Prowlarr/Prowlarr/blob/develop/src/NzbDrone.Core/Indexers/NewznabStandardCategory.cs

#### Docs/Wiki

- [x] [Wiki Updates](https://wiki.servarr.com)
- [x] [Cardigann YML Definition Updates](https://wiki.servarr.com/en/prowlarr/cardigann-yml-definition)
    - [x] TMDBId
    - [x] Genre
    - [x] TraktID
    - [x] CategoryDescr
    - [x] Protocol
    - [x] ignoreblankinputs (aka SupportsEmptyInputs)
    - [x] allowdownloadredirect (defined as SupportsRedirect)
    - [x] SearchBlock Error
    - [x] DownloadBlock Error

#### Issues Fixed or Closed by this PR

* Fixes #261
* Towards #259 
* Fixes #782
* Fixes #680 
* Closes #681 
* Closes #751
* Fixes #807 

### Prior to Merge ToDo
- [x] Change Definition Branch back to `master`

### CODE REVIEW

File: src/NzbDrone.Core/Indexers/Definitions/Cardigann/CardigannBase.cs [hunk 1]
```diff
@@ -65,38 +65,54 @@ public class CardigannBase
             {
                 foreach (var category in _definition.Caps.Categories)
                 {
-                    var cat = NewznabStandardCategory.GetCatByName(category.Value);
-                    if (cat == null)
+                    if (_definition.Protocoltype == "usenet")
                     {
-                        _logger.Error(string.Format("CardigannIndexer ({0}): invalid Torznab category for id {1}: {2}", _definition.Id, category.Key, category.Value));
-                        continue;
+                        var cat = category;
```
Comment (bakerboy448) [inline src/NzbDrone.Core/Indexers/Definitions/Cardigann/CardigannBase.cs:70]: needs work

File: src/NzbDrone.Core/Indexers/Definitions/Cardigann/CardigannBase.cs [hunk 2]
```diff
@@ -65,38 +65,54 @@ public class CardigannBase
...
                 foreach (var categorymapping in _definition.Caps.Categorymappings)
                 {
-                    IndexerCategory torznabCat = null;
-
-                    if (categorymapping.cat != null)
+                    if (_definition.Protocoltype == "usenet")
                     {
-                        torznabCat = NewznabStandardCategory.GetCatByName(categorymapping.cat);
-                        if (torznabCat == null)
+                        int.TryParse(categorymapping.id, out var categorymappingidint);
```
Comment (bakerboy448) [inline src/NzbDrone.Core/Indexers/Definitions/Cardigann/CardigannBase.cs:93]: needs work

do we want the YML to feed categories or category mapping?

File: src/NzbDrone.Core/Indexers/Definitions/Cardigann/CardigannParser.cs [hunk 3]
```diff
@@ -436,16 +438,19 @@ private string ParseFields(string value, string fieldName, TorrentInfo release,
                     value = release.Description;
                     break;
                 case "category":
-                    var cats = MapTrackerCatToNewznab(value);
-                    if (cats.Any())
+                    if (protocoltype == "torrents")
```
Comment (bakerboy448) [inline src/NzbDrone.Core/Indexers/Definitions/Cardigann/CardigannParser.cs:441]: need to handle usenet cats here

File: src/NzbDrone.Core/Indexers/Definitions/Cardigann/Cardigann.cs [hunk 4]
```diff
@@ -127,11 +127,11 @@ private IndexerDefinition GetDefinition(CardigannMetaDefinition definition)
                 Implementation = GetType().Name,
                 IndexerUrls = definition.Links.ToArray(),
                 Settings = new CardigannSettings { DefinitionFile = definition.File },
-                Protocol = DownloadProtocol.Torrent,
+                Protocol = definition.Protocoltype == "usenet" ? DownloadProtocol.Usenet : DownloadProtocol.Torrent,
```
Comment (bakerboy448) [inline src/NzbDrone.Core/Indexers/Definitions/Cardigann/Cardigann.cs:130]: Need to adjust the Download Call in Cardigann to be conditional based on the protocol

**Comment by bakerboy448** [comment]
see OP / need some help with:
- stuck on the selectors for categories, etc.
- downloads end up as {title}.torrents
- YML defs updated; DS added in addition to geek

also need help with:
- Cardigann is SLOWWWWWWW

positive note:
Category mapping for YML does work now as expected. e.g. Web-DL searches Slug's category rather than mismatching category numbers

**Comment by bakerboy448** [comment]
ToDo: update definitions https://github.com/Prowlarr/Prowlarr/commit/579b8a3d3bdb293fdbf2dbbdbaf9eafdef4e40ef

File: src/NzbDrone.Core/Indexers/Definitions/Cardigann/CardigannDefinition.cs [hunk 5]
```diff
@@ -40,6 +40,8 @@ public class CardigannDefinition
         public List<string> Legacylinks { get; set; }
         public bool Followredirect { get; set; } = false;
         public bool TestLinkTorrent { get; set; } = true;
+        public string Protocoltype { get; set; }
```
Comment (Qstick) [inline src/NzbDrone.Core/Indexers/Definitions/Cardigann/CardigannDefinition.cs:43]: `Protocol`

**Comment by bakerboy448** [comment]
To Fix:  Queries of non base categories not using the category

**Comment by bakerboy448** [comment]
ToDo - Change to Categories (See Jackett fbd48823045d58d282ab5ec7498ce6a9f3cb8280)
ToDo - Ensure we support Categories, @Qstick  - I'd assume we do from Cardigann? or was that left out :D

**Comment by bakerboy448** [comment]
Next Steps
- [ ] - Add Search/Download Error Handling (newznab errors)
- Discuss how/if to `obsolete` C# Newznab
- [ ] - Once errors are handled - Time for Testing

File: src/NzbDrone.Core/IndexerVersions/IndexerDefinitionUpdateService.cs [hunk 6]
```diff
@@ -25,8 +25,10 @@ public interface IIndexerDefinitionUpdateService
     public class IndexerDefinitionUpdateService : IIndexerDefinitionUpdateService, IExecute<IndexerDefinitionUpdateCommand>
     {
         /* Update Service will fall back if version # does not exist for an indexer  per Ta */
+        private const int DEFINITION_VERSION = 4;
 
-        private const int DEFINITION_VERSION = 3;
+        // ToDo Update branch back to master
+        private const string DEFINITION_BRANCH = "usenet-v4";
```
Comment (bakerboy448) [inline src/NzbDrone.Core/IndexerVersions/IndexerDefinitionUpdateService.cs:31]: leaving this open as a very obvious todo reminder - change back to master

File: src/NzbDrone.Core/Indexers/Definitions/Cardigann/Cardigann.cs [hunk 7]
```diff
@@ -136,7 +136,7 @@ private IndexerDefinition GetDefinition(CardigannMetaDefinition definition)
                 },
                 SupportsRss = SupportsRss,
                 SupportsSearch = SupportsSearch,
-                SupportsRedirect = SupportsRedirect,
+                SupportsRedirect = definition.Allowownloadredirect,
```
Comment (bakerboy448) [inline src/NzbDrone.Core/Indexers/Definitions/Cardigann/Cardigann.cs:139]:
```suggestion
                SupportsRedirect = definition.Allowdownloadredirect,
```

File: src/NzbDrone.Core/Indexers/Definitions/Cardigann/CardigannMetaDef.cs [hunk 8]
```diff
@@ -16,5 +16,7 @@ public class CardigannMetaDefinition
         public List<SettingsField> Settings { get; set; }
         public string Sha { get; set; }
         public LoginBlock Login { get; set; }
+        public string Protocol { get; set; }
+        public bool Allowownloadredirect { get; set; }
```
Comment (bakerboy448) [inline src/NzbDrone.Core/Indexers/Definitions/Cardigann/CardigannMetaDef.cs:20]:
```suggestion
        public bool Allowdownloadredirect { get; set; }
```

File: src/NzbDrone.Core/Indexers/Definitions/Cardigann/CardigannParser.cs [hunk 9]
```diff
@@ -358,27 +361,30 @@ public IList<ReleaseInfo> ParseResponse(IndexerResponse indexerResponse)
...
+            if (protocol == "torrent")
             {
-                // generate magnet link from info hash (not allowed for private sites)
-                if (((TorrentInfo)c).MagnetUrl == null && !string.IsNullOrWhiteSpace(((TorrentInfo)c).InfoHash) && _definition.Type != "private")
+                releases.ForEach(c =>
                 {
-                    ((TorrentInfo)c).MagnetUrl = MagnetLinkBuilder.BuildPublicMagnetLink(((TorrentInfo)c).InfoHash, c.Title);
-                }
+                    // generate magnet link from info hash (not allowed for private sites)
+                    if (((TorrentInfo)c).MagnetUrl == null && !string.IsNullOrWhiteSpace(((TorrentInfo)c).InfoHash) && _definition.Type != "private")
```
Comment (bakerboy448) [inline src/NzbDrone.Core/Indexers/Definitions/Cardigann/CardigannParser.cs:369]: should we also exclude semi-private?

just change to `= public`

**Comment by bakerboy448** [comment]
wiki updated https://wiki.servarr.com/en/prowlarr/cardigann-yml-definition

https://github.com/Servarr/Wiki/commit/9734880ec756400f6321a7f6ad5628d8b99cb7d1
https://github.com/Servarr/Wiki/commit/1668937c05934c09cdc79022bc8e4d6bd996fa09

File: src/NzbDrone.Core/Indexers/Definitions/Cardigann/CardigannParser.cs (hunk 9 above)
Comment (bakerboy448) [reply src/NzbDrone.Core/Indexers/Definitions/Cardigann/CardigannParser.cs:369]: @Qstick  bump

**Comment by bakerboy448** [comment]
- [ ] - @Qstick did we check if cardigann correctly handles indexer errors for usenet? E.g. query limit reached etc and back off?

File: src/NzbDrone.Core/Indexers/Definitions/Cardigann/CardigannBase.cs [hunk 10]
```diff
@@ -816,5 +837,57 @@ protected Uri ResolvePath(string path, Uri currentUrl = null)
...
+            {
+                case "html":
+                    resultParser = new HtmlParser();
+                    resultDocument = resultParser.ParseDocument(response.Content);
+                    break;
+                case "xml":
+                    resultParser = new XmlParser();
+                    resultDocument = resultParser.ParseDocument(response.Content);
+                    break;
+                case "json":
```
Comment (bakerboy448) [inline src/NzbDrone.Core/Indexers/Definitions/Cardigann/CardigannBase.cs:868]: We'd want something for json  for BTN and PTP plus other Json responses no?

File: src/NzbDrone.Core/Indexers/Definitions/Cardigann/CardigannParser.cs (hunk 9 above)
Comment (bakerboy448) [inline src/NzbDrone.Core/Indexers/Definitions/Cardigann/CardigannParser.cs:369]:
```suggestion
                    if (((TorrentInfo)c).MagnetUrl == null && !string.IsNullOrWhiteSpace(((TorrentInfo)c).InfoHash) && _definition.Type = "pubic")
```

**Comment by bakerboy448** [comment]
Broken Piece - Testing DrunkenSlug Fails

```
[Trace] IndexerHttpClient: Res: [GET] https://drunkenslug.com/api/?t=search&apikey=(removed)&q=&rid=&tmdbid=&tvdbid=&imdbid=&traktid=&season=&ep=&album=&artist=&label=&genre=&year=&cat=&extended=1: 201.Created (218 ms)
[Debug] Cardigann: Parsing
[Trace] HttpClient: Res: [GET] https://prowlarr.servarr.com/v1/time: 200.OK (120 ms)
[Trace] ConfigService: Using default config value for 'logindexerresponse' defaultValue:'False'
[Trace] Cardigann: Unexpected Response content (115 bytes): <?xml version="1.0" encoding="UTF-8"?>
<error code="201" description="Incorrect parameter (q must not be empty)"/>

[Trace] ConfigService: Using default config value for 'uilanguage' defaultValue:'1'
```

**Comment by Qstick** [comment]
Feel free to keep this open for reference for now, but can we reduce v4 to these commits. We'll handle usenet in v5 with #823. We'll also need to dial back the newznab defs that have gone into v4 indexers repo

image

**Comment by bakerboy448** [comment]
@Qstick you sure we don't want the searchblock and downblock? I thought those are in YML now and we ignore them?

**Comment by bakerboy448** [comment]
closing; replaced elsewhere

will keep branch until Cardigann v4 lands

**Pull request closed without merging**